### PR TITLE
consensus: add Istanbul BFT

### PR DESCRIPTION
**Pull request by yutelin**
This PR is the implementation of Istanbul BFT protocol in ethereum/EIPs#650

### CODE REVIEW

**Comment by joelburget** [comment]
If you could rebase to the latest geth1.6 that would be great. The merge conflicts are very easy.

I plan to test using quorum-tools on [this branch](https://github.com/jpmorganchase/quorum-tools/tree/feature/istanbul).

**Comment by yutelin** [comment]
@joelburget  done.

**Comment by patrickmn** [comment]
This is really great work!

Please execute the CLA which is required for merging: https://github.com/jpmorganchase/cla

**Comment by yutelin** [comment]
Just done with latest rebasing and sending CLA. @patrickmn

**Comment by patrickmn** [comment]
Received--thank you!

File: cmd/utils/flags.go [hunk 1]
```diff
@@ -124,7 +124,7 @@ var (
 	}
 	NetworkIdFlag = cli.Uint64Flag{
 		Name:  "networkid",
-		Usage: "Network identifier (integer, 1=Frontier, 2=Morden (disused), 3=Ropsten, 4=Rinkeby)",
+		Usage: "Network identifier (integer, 1=Frontier, 2=Morden (disused), 3=Ropsten, 4=Rinkeby, 5=Ottoman)",
```
Comment (bts) [inline cmd/utils/flags.go:127]: Seems like we might want to use another (higher) number that's less likely to cause a conflict in the future?

Comment (yutelin) [reply cmd/utils/flags.go:127]: It depends on the upstream codebase. We submit the same code to geth, so if that gets merged, we are going to grab network id 5. If we eventually get a larger number, I think we can update it here accordingly.

**Comment by yutelin** [comment]
We've created a PR review guide document.
https://github.com/getamis/quorum/wiki/PR-Review-Guide
Please see if that helps.

File: consensus/consensus.go [hunk 2]
```diff
@@ -90,6 +91,21 @@ type Engine interface {
 
 	// APIs returns the RPC APIs this consensus engine provides.
 	APIs(chain ChainReader) []rpc.API
+
+	// Protocol returns the protocol for this consensus
+	Protocol() Protocol
+}
+
+// Handler should be implemented is the consensus needs to handle and send peer's message
```
Comment (armaniferrante) [inline consensus/consensus.go:99]: Typo? Should "is" be "if"?

**Comment by patrickmn** [comment]
Ok to close this?

**Comment by AUQIB92** [comment]
@yutelin 
 
 please help
I m getting this error
Block sealing failed                     err=unauthorized
ERROR[07-17|22:08:43] Failed to decode message from payload    address=0x80A13E1aa941569081c696525e38B6a65d1c8603 err="unauthorized address"

My environment

Geth
Version: 1.8.3-unstable
Git Commit: c7547381b2ea8999e423970d619835c662176790
Architecture: amd64
Network Id: 1
Go Version: go1.10.4
Operating System: linux ubuntu 16.04
GOPATH=
GOROOT=/usr/lib/go-1.10

**Comment by prd-fox** [comment]
@AUQIB92 please don't commit on closed issues. Please open a new issue with your problem and one of the team will look into it.

**Pull request closed without merging**